### PR TITLE
Changed all strcpy to pbs_strncpy in scheduler

### PR DESCRIPTION
**Pull request by arungrover**
#### Describe Your Change
Using strcpy at times can be risky and cause problems like buffer overrun.

In this PR, I've added a generic pbs_strncpy function that always puts a '\0' at the end of the buffer.

Changed all strcpy in the scheduler module to pbs_strncpy.

In future, we need to also make changes to other modules.

#### Link to Design Doc
N/A

#### Attach Test and Valgrind Logs/Output

[smoketest.txt](https://github.com/openpbs/openpbs/files/4974728/smoketest.txt)

### CODE REVIEW

**Comment by agrawalravi90** [comment]
@arungrover  just curious, are we adding this because snprintf() can be slow?

File: src/lib/Libutil/misc_utils.c [hunk 1]
```diff
@@ -334,6 +334,26 @@ pbs_strcpy(char *dest, const char *src)
...
+ *
+ * @return char *
+ * @retval pointer to the destination string
+ *
+ * @note: Caller needs ensure non-NULL pointers
+ */
+char *pbs_strncpy(char *dest, const char *src, size_t n)
+{
+    strncpy(dest, src, n-1);
+    dest[n-1] = '\0';
```
Comment (bhroam) [inline src/lib/Libutil/misc_utils.c:354]: Add spaces around the '-'

File: src/scheduler/fairshare.c [hunk 2]
```diff
@@ -627,7 +627,7 @@ write_usage(char *filename, fairshare_head *fhead)
 	 * ...
 	 */
 
-	strcpy(head.tag, USAGE_MAGIC);
+	pbs_strncpy(head.tag, USAGE_MAGIC, 9);
```
Comment (bhroam) [inline src/scheduler/fairshare.c:630]: You use a combination of static numbers and macros for the size of pbs_strncpy().  Just use sizeof(dest) everywhere.

File: src/scheduler/parse.c [hunk 3]
```diff
@@ -103,7 +103,7 @@ parse_config(char *fname)
 	FILE *fp;			/* file pointer to config file */
 	char *buf = NULL;
 	int buf_size = 0;
-	char buf2[8192];		/* general purpose buffer */
+	char buf2[PBS_FGETS_LINE_LEN];	/* general purpose buffer */
```
Comment (bhroam) [inline src/scheduler/parse.c:106]: Why PBS_FGETS_LINE_LEN?  This has nothing to do with fgets.

Comment (arungrover) [reply src/scheduler/parse.c:106]: You are right, this buffer isn't used with fgets. Reverting it.

File: src/scheduler/fairshare.c (hunk 2 above)
Comment (arungrover) [reply src/scheduler/fairshare.c:630]: True, It didn't occur to me :)
I'll change it.

File: src/scheduler/pbs_sched.c [hunk 4]
```diff
@@ -732,7 +732,7 @@ lock_out(int fds, int op)
 	flock.l_start  = 0;
 	flock.l_len    = 0;	/* whole file */
 	if (fcntl(fds, F_SETLK, &flock) < 0) {
-		pbs_strncpy(log_buffer, "pbs_sched: another scheduler running\n", LOG_BUF_SIZE);
+		pbs_strncpy(log_buffer, "pbs_sched: another scheduler running\n", sizeof(log_buffer));
 		log_err(errno, msg_daemonname, log_buffer);
```
Comment (bhroam) [inline src/scheduler/pbs_sched.c:736]: This is a static string, mind skipping the strcpy() and moving it directly into the log_err()?

Comment (arungrover) [reply src/scheduler/pbs_sched.c]: I've checked in all places, there isn't an instance where I'm doing sizeof on a pointer pointing to dynamically allocated memory.

**Comment by arungrover** [comment]
sorry, I missed that @agrawalravi90 
Yes, I ran some tests and found that snprintf can be many times slower than strncpy. In one of the test, strncpy took 1 second to copy a small string 100 million times and snprintf took 10 seconds.

**Comment by vstumpf** [comment]
We have a bunch of strncpys, shoudl we change those too?

**Comment by arungrover** [comment]
@vstumpf while I agree with you that all calls should use one function, this particular PR was to address issues with strcpy only. Maybe we can change strncpys as and when we encounter them in future work.

**Comment by arungrover** [comment]
@agrawalravi90 I've changed the description.

**Pull request merged**